staging/src/k8s.io/apiserver/pkg/storage/etcd3/errors.go

```diff
@@ -30,13 +32,26 @@ func interpretWatchError(err error) error {
 	return err
 }
 
-func interpretListError(err error, paging bool) error {
-	switch {
-	case err == etcdrpc.ErrCompacted:
-		if paging {
-			return errors.NewResourceExpired("The provided from parameter is too old to display a consistent list result. You must start a new list without the from.")
-		}
+func interpretListError(err error, continueKey, keyPrefix string) error {
+	if err != etcdrpc.ErrCompacted {
+		return err
+	}
+	if len(continueKey) == 0 {
 		return errors.NewResourceExpired("The resourceVersion for the provided list is too old.")
 	}
+	message := "The provided from parameter is too old to display a consistent list result. " +
+		"You can start a new list without the from."
+	// continueToken.ResoureVersion=-1 means that the apiserver can
+	// continue the list at the latest resource version. We don't use rv=0
+	// for this purpose to distinguish from a bad token that has empty rv.
+	newToken, err := encodeContinue(continueKey, keyPrefix, -1)
+	if err != nil {
+		return errors.NewResourceExpired(message +
+			fmt.Sprintf("The apiserver cannot provide a continue token for inconsistent list: %s", err))
+	}
+	err = errors.NewResourceExpired(message +
+		"Or use the continue token in this response to continue the list, " +
+		"though the next chunk will be from a different snapshot than the previous chunks.")
```

Comment:
I'm not sure a regular user will understand what "different snapshot" means.  Try to find something more clear.